traitsui/tests/_tools.py

```diff
@@ -24,48 +25,86 @@
+    collected.
     """
+    return (
+        str(exc_type),
+        str(value),
+        str("".join(traceback.format_exception(exc_type, value, tb)))
+    )
+
+
+@contextmanager
+def reraise_exceptions(logger=_TRAITSUI_LOGGER):
+    """ Context manager to capture all exceptions occurred in the context and
+    then reraise a RuntimeError if there are any exceptions captured.
 
-    exceptions = []
+    Exceptions from traits change notifications are also captured and reraised.
 
-    def _print_uncaught_exception(type, value, tb):
-        message = "Uncaught exception:\n"
-        message += "".join(traceback.format_exception(type, value, tb))
-        print(message)
+    Depending on the GUI toolkit backend, unexpected exceptions occurred in the
+    GUI event loop may (1) cause fatal early exit of the test suite or (2) be
+    printed to the console without causing the test to error. This context
+    manager is intended for testing purpose such that unexpected exceptions
+    will result in a test error.
+
+    Parameters
+    ----------
+    logger : logging.Logger
+        Logger to use for logging errors.
+    """
+    serialized_exceptions = []
 
     def excepthook(type, value, tb):
-        exceptions.append(value)
-        _print_uncaught_exception(type, value, tb)
+        serialized = _serialize_exception(type, value, tb)
+        serialized_exceptions.append(serialized)
+        logger.error(
+            "Unexpected error captured in sys excepthook. \n%s",
+            serialized[-1]
+        )
 
-    def handle_exception(object, trait_name, old, new):
+    def handler(object, name, old, new):
         type, value, tb = sys.exc_info()
-        exceptions.append(value)
-        _print_uncaught_exception(type, value, tb)
+        serialized_exceptions.append(_serialize_exception(type, value, tb))
+        logger.exception(
+            "Unexpected error occurred from change handler "
+            "(object: %r, name: %r, old: %r, new: %r).",
+            object, name, old, new,
+        )
 
-    _original_handle_exception = traits.trait_notifiers.handle_exception
+    push_exception_handler(handler=handler)
+    sys.excepthook = excepthook
     try:
-        sys.excepthook = excepthook
-        traits.trait_notifiers.handle_exception = handle_exception
         yield
     finally:
         sys.excepthook = sys.__excepthook__
-        traits.trait_notifiers.handle_exception = _original_handle_exception
-        if len(exceptions) > 0:
-            raise exceptions[0]
+        pop_exception_handler()
+        if serialized_exceptions:
+            msg = "Uncaught exceptions found.\n"
+            msg += "\n".join(
+                "=== Exception (type: {}, value: {}) ===\n"
+                "{}".format(*record)
+                for record in serialized_exceptions
+            )
+            raise RuntimeError(msg)
+
+
+# Temporary alias to avoid conflicts across PRs.
+store_exceptions_on_all_threads = reraise_exceptions
```

Comment:
If this is only used internally (which i think is the case), we can just remove the alias once all of it's callers/users have been updated - preferably in a separate PR.

Reply:
for now, we could just have an open issue for this change.

Reply:
I can do that renaming in a separate PR.
(The renaming is a drive-by too... the original name is a bit confusing: The motivation of this context manager has nothing to do with threading.)

Reply:
Done: https://github.com/enthought/traitsui/pull/1102